sql: panic: runtime error: invalid memory address or nil pointer dereference

```
SELECT json_agg((3808362714,));
```

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: panic while executing 1 statements: SELECT json_agg((_,)); caused by runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x60 pc=0x131c42a]

goroutine 590 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc001884d80, 0x456eec0, 0xc0006def80, 0x3870500, 0x6613eb0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:727 +0x330
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1(0xc001884d80, 0x456eec0, 0xc0006def80)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:440 +0x61
panic(0x3870500, 0x6613eb0)
	/usr/local/go/src/runtime/panic.go:522 +0x1b5
github.com/cockroachdb/cockroach/pkg/sql/types.(*T).TupleLabels(...)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/types/types.go:799
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.AsJSON(0x45cfec0, 0xc001649180, 0x7fbf05215460, 0x0, 0xc000346060, 0xc0017ef950)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/datum.go:2528 +0x5da
github.com/cockroachdb/cockroach/pkg/sql/sem/builtins.(*jsonAggregate).Add(0xc0016dc2d0, 0x456eec0, 0xc00165a1c0, 0x45cfec0, 0xc001649180, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/builtins/aggregate_builtins.go:2079 +0x39
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*aggregatorBase).accumulateRowIntoBucket(0xc0011bac00, 0xc0016dc2a0, 0x1, 0x1, 0x0, 0x0, 0x0, 0xc0018a9030, 0x1, 0x1, ...)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/aggregator.go:798 +0x579
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*orderedAggregator).accumulateRow(0xc0011bac00, 0xc0016dc2a0, 0x1, 0x1, 0xc0016dc2a0, 0x1)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/aggregator.go:851 +0xbe
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*orderedAggregator).accumulateRows(0xc0011bac00, 0x396c040, 0xc0003dd300, 0xc00165a1c0, 0x3cce395, 0x12)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/aggregator.go:557 +0x233
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*orderedAggregator).Next(0xc0011bac00, 0xc00165a1c0, 0x3cce395, 0x12, 0x456eec0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/aggregator.go:726 +0x14d
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.Run(0x456eec0, 0xc00165a1c0, 0x4581a40, 0xc0011bac00, 0x454bc00, 0xc00184f880)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/base.go:171 +0x35
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*ProcessorBase).Run(0xc0011bac00, 0x456eec0, 0xc00165a1c0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/processors.go:793 +0x92
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).Run(0xc0018f2000, 0x456ef80, 0xc0016dc180, 0x3e30200, 0x0, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:584 +0x1ec
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc0008185a0, 0xc0016f82a0, 0xc0017ef440, 0xc0015b2870, 0xc00073c000, 0xc0018851a8, 0x0, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:333 +0x381
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun(0xc0008185a0, 0x456ef80, 0xc00166fda0, 0xc0018851a8, 0xc0016f82a0, 0xc0017ef440, 0x45708c0, 0xc0002b3200, 0xc00073c000, 0xc001884f58)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:942 +0x20a
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine(0xc001884d80, 0x456ef80, 0xc00166fda0, 0xc0018850d0, 0x3, 0x7fbf01bd65d0, 0xc00151e668, 0xc0015b2a01, 0x0, 0x0, ...)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:826 +0x370
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc001884d80, 0x456ef80, 0xc00166fda0, 0xc0018850d0, 0x7fbf01bd65d0, 0xc00151e668, 0x0, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:705 +0x6e0
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc001884d80, 0x456ef80, 0xc00166fda0, 0x457b180, 0xc0015a9f40, 0xc000ae00e6, 0x1e, 0x0, 0x0, 0x0, ...)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:417 +0xb64
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc001884d80, 0x456ef80, 0xc00166fda0, 0x457b180, 0xc0015a9f40, 0xc000ae00e6, 0x1e, 0x0, 0x0, 0x0, ...)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:99 +0x4ec
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0xc001884d80, 0x456eec0, 0xc0015a9f80, 0x0, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1211 +0x1aba
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc001884d80, 0x456eec0, 0xc0006def80, 0xc000532e80, 0x5400, 0x15000, 0xc000532f18, 0xc000424300, 0x0, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1140 +0x1a3
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc0006fcd80, 0x456eec0, 0xc0006def80, 0xc001884d80, 0x5400, 0x15000, 0xc000532f18, 0xc000424300, 0x0, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:442 +0xce
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1(0xc0002cf3f9, 0xc0002fa6e0, 0x456eec0, 0xc0006def80, 0xc000424300, 0xc0006fcd80, 0xc00151e000, 0x4579700, 0xc0002fa6a0, 0xc000fc6240, ...)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:584 +0x21c
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:513 +0x17b
```

--- a/pkg/sql/logictest/testdata/logic_test/json_builtins
+++ b/pkg/sql/logictest/testdata/logic_test/json_builtins
@@ -327,6 +327,10 @@ SELECT to_json(x.*) FROM (VALUES (1,2)) AS x(a);
 query error found duplicate tuple label: \"column2\"
 SELECT to_json(x.*) FROM (VALUES (1,2)) AS x(column2);
 
+# Regression test for #39502.
+statement ok
+SELECT json_agg((3808362714,))
+
 ## json_array_elements and jsonb_array_elements
 
 query T

--- a/pkg/sql/sem/tree/datum.go
+++ b/pkg/sql/sem/tree/datum.go
@@ -2525,6 +2525,10 @@ func AsJSON(d Datum) (json.JSON, error) {
 		return builder.Build(), nil
 	case *DTuple:
 		builder := json.NewObjectBuilder(len(t.D))
+		// We need to make sure that t.typ is initialized before getting the tuple
+		// labels (it is valid for t.typ be left uninitialized when instantiating a
+		// DTuple).
+		t.maybePopulateType()
 		labels := t.typ.TupleLabels()
 		for i, e := range t.D {
 			j, err := AsJSON(e)
@@ -2681,8 +2685,9 @@ func AsDTuple(e Expr) (*DTuple, bool) {
 	return nil, false
 }
 
-// ResolvedType implements the TypedExpr interface.
-func (d *DTuple) ResolvedType() *types.T {
+// maybePopulateType populates the tuple's type if it hasn't yet been
+// populated.
+func (d *DTuple) maybePopulateType() {
 	if d.typ == nil {
 		contents := make([]types.T, len(d.D))
 		for i, v := range d.D {
@@ -2690,6 +2695,11 @@ func (d *DTuple) ResolvedType() *types.T {
 		}
 		d.typ = types.MakeTuple(contents)
 	}
+}
+
+// ResolvedType implements the TypedExpr interface.
+func (d *DTuple) ResolvedType() *types.T {
+	d.maybePopulateType()
 	return d.typ
 }
 

